### ISSUE TITLE
LocalCluster memory usage and inspection

### ISSUE THREAD

**Issue by aavanian** [dask/distributed#1834]
I started writing that in gitter but I think it was running too long and didn't seem fit for SO.

I run a local cluster with 4 workers which prepare, persist and publish a dataset and a few clients query/perform computations on during the day. I upgraded recently from dask 0.15.4 to 0.17.1 and distributed from 1.19.1 to 1.21.1 and the cluster is behaving quite differently: worker memory is very volatile (published dataset is about 2+gB but dashboard shows anything between 150mB and 6gB) and processing is much slower.

Superficial observation and a bit of speculation makes me think the memory usage (or memory limit) has significantly increased (not from the data which grows very slowly every day) and the workers which never or rarely dumped to disk before now do so for almost any new computation.

A few questions from there:
* Is there a way to inspect which data a worker currently has in memory and in particular if the published dataset has been dumped to disk (which I'd like to avoid)?
* once the dataset is persisted, there is a priori no operation that should lead to a transfer of its partitions between workers except for gathering results at the end of a computation (and I guess at this stage it should be a copy of a subset of the dataset, not itself). Yet it seems that the worker memory load becomes unbalanced as time and computations goes by (that being said, data is possibly resting on disk and the worker shows light memory load because of that).

**Comment by mrocklin** [comment]
> Is there a way to inspect which data a worker currently has in memory and in particular if the published dataset has been dumped to disk (which I'd like to avoid)?

There are a few ways.  One way would be to start your local cluster with `processes=False` and then inpsect the worker state directly

```python
client = Client(processes=False, n_workers=4, threads_per_worker=1)
# do work
client.cluster.workers[0].data.fast # in-memory data
client.cluster.workers[0].data.slow # on-disk data
```

Some of this is also available on the dashboard

> once the dataset is persisted, there is a priori no operation that should lead to a transfer of its partitions between workers except for gathering results at the end of a computation (and I guess at this stage it should be a copy of a subset of the dataset, not itself). Yet it seems that the worker memory load becomes unbalanced as time and computations goes by (that being said, data is possibly resting on disk and the worker shows light memory load because of that).

Correct, I would not expect any transfer of data if you're not actively computing.

I apologize for the delay in response.  I was out last week and relatively few others visit this part of the codebase.

**Comment by mrocklin** [comment]
Any update on this @aavanian ?

**Comment by aavanian** [comment]
I'm closing this issue, it will remain partly a mystery.
Workers started to quit on "out of memory" after the dask+distributed upgrade but a revert to previous version didn't fix the issue.
I investigated various ideas but couldn't find something concrete. I think there's some memory leak at the worker level but it could be the profile data? In any case it's smaller than what caused the issue.

In the process, I realized that my data is actually bigger in memory than reported by pandas (there are nested objects but even with `deep=True`, it's underestimating it) and fixed my issue by removing some of the nested data + increased number of partitions.
